### ISSUE TITLE
eventual goal: re-remove leak-check from compiler 

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#59490]
PR #58592 re-added the so-called leak check.

Re-adding the leak-check masked a number of bugs, where you need to use `-Z no-leak-check` to observe them again.

We plan to eventually re-remove the leak check. But before doing so, we need to double-check that all of the aforementioned masked bugs are either fixed, or at the very least revisited in terms of determining their priority in a leak-check-free world.

**Comment by vi** [comment]
What is a leak check? How it masks the bugs? What Github label do that leak-check-masked bugs carry?

Google search does not seem to show relevant things; asking on IRC lead to answeres suggesting to ask here.

Do I understand the chain correctly:

1. fixing those bugs unlocks
2. removing re-added leak-check, which unlocks
3. Chalk integration, which unlocks
4. generic associated types, which unlocks
5. better abstraction abilities for Rust, which unlocks
6. better and safer APIs.

**Comment by mark-i-m** [comment]
@vi perhaps @pnkfelix can correct me if i am mistaken. I think they are referring to [Placeholder leaks](https://rust-lang.github.io/rustc-guide/traits/hrtb.html?highlight=Leak#basic-matching-and-placeholder-leaks), which was the old way of handling HRTB lifetimes. My guess is that there is some difference in how Universes (the new way) work that caused edge case bugs or something....

**Comment by pnkfelix** [comment]
> I think they are referring to [Placeholder leaks](https://rust-lang.github.io/rustc-guide/traits/hrtb.html?highlight=Leak#basic-matching-and-placeholder-leaks), which was the old way of handling HRTB lifetimes.

Yes, this matches my own understanding.

The check that was re-added in PR #58592 has a big comment above its core function, here:

https://github.com/rust-lang/rust/blob/33d3598e3b57c0fcfd36f34c346847bcc4edd3d2/src/librustc/infer/region_constraints/leak_check.rs#L7-L30


I think the [rustc guide mark-i-m linked](https://rust-lang.github.io/rustc-guide/traits/hrtb.html?highlight=Leak#basic-matching-and-placeholder-leaks) is the best place for a new-comer to start; but I also do not know if the strategy outlined for handling leaks corresponds to our ideal-but-not-yet-the-default solution, or if the ideal solution is not yet documented there.

----

I am not an expert here. PR #58592 says the main reason to re-remove the leak check is that the leak-check does not handle higher-ranked types correctly in all cases.

What I do not know is whether the remaining buggy cases motivating removal of the leak check are all instances of expressiveness bugs (programs that we want to accept but currently reject), or if some of them are soundness issues. I *think* it is just an expressiveness issue, but I am not sure of that.

**Comment by pnkfelix** [comment]
> What is a leak check? How it masks the bugs? What Github label do those leak-check-masked bugs carry?

I say that the leak check is "masking bugs" because it is causing the `rustc` compiler to reject programs relatively early (due to them violating the leak check), but those same programs, if we do get rid of the leak check, end up hitting some bug downstream that causes an internal compiler error.

Currently all such bugs that I know of are things that arise from interactions with non-lexical lifetimes, perhaps in part due to the MIR typeck pass that is in charge of checking the correctness of higher-ranked types (and without the leak check, a larger class of potential higher-ranked types ends up being fed into the MIR typeck pass).

It is *possible* that there are other bugs that belong in this category that are not due to interaction with non-lexical lifetime. I am just not yet aware of any such cases; it is an unknown quantity.

We have not allocated a github-label dedicated soley to these masked bugs (and I do not think that we should allocate such a label). I think it will suffice for such bugs to mention this isue (#59490) so that they end up being linked from this page.

**Comment by nikomatsakis** [comment]
So #65232 made good progress towards this goal by adding universes to the `'empty` regions and thus ensuring that we don't have any region `R1` where `forall<R2> { R2: R1 }` is true. However, this work did not extend to the NLL checker. Therefore, I think the major remaining work items are as follows:

* extend the NLL check to avoid a notion of a `'empty` value
* resolve the [confusing error messages](https://github.com/rust-lang/rust/pull/65232#issuecomment-540215930) that may result
* integrate empty into the canonicalizer ([see FIXME](https://github.com/rust-lang/rust/blob/94fe66d3df05a5f6329bae64e02dfe5ba09eb71a/src/librustc/infer/canonical/canonicalizer.rs#L202-L204))
* figure out how to resolve the [wasm-bindgen](https://github.com/rust-lang/rust/pull/65232#issuecomment-583116754) regression

**Comment by nikomatsakis** [comment]
I'm looking some into that first point and will probably create a spin-out issue for it and put more details there.

**Comment by lcnr** [comment]
discussed in the [t-types backlog bonanza](https://rust-lang.zulipchat.com/#narrow/stream/326132-t-types.2Fmeetings/topic/meeting.202022-07-15/near/289724216)

While the plan for *how* we want to remove the leak check has changed, we still intend to remove the current version, so we're keeping this open to track that for now.

The goal is to pretty much add a `param_env` aware leak check directly into trait solving and remove the current impl.

**Comment by pnkfelix** [comment]
Discussed in [T-compiler backlog bonanza](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bsteering.20meeting.5D.202022-08-12.20compiler-team.23484.20backlog.20b'za/near/293115078)

@rustbot label: S-tracking-unimplemented

**Comment by lcnr** [comment]
I don't think this tracking issue is useful. We are reworking the way region handling works which is tracked in https://github.com/rust-lang/rust-project-goals/blob/main/src/2026/assumptions_on_binders.md. I don't think this tracking issue is useful for this and think the project goal will do a significantly better job here
